text/0091-ci-upload-tokens.md

```diff
@@ -0,0 +1,271 @@
+not put a user on the path of success but requires multiple separate steps to get
+everything in order.
+
+# Background
+
+We improved a lot of the inner workings of source maps and debug files at Sentry recently
+but these efforts are held back by the complexity of getting the token.  The friction is
+still too high for many customers to make the necessary investment into getting source maps
+uploaded.  From a documentation writing and onboarding experience, it's also not clear with
+the current system how the experience can be improved.
+
+Additionally both Hybrid Cloud and Single Tenant would greatly benefit from automatically
+routing to the right URLs.  Today the documentation is very quiet about how to get this
+system to work on a single tenant installation and customers are often required to work
+with CS to get source maps working.
+
+# Technical Implementation
+
+The motivation is to add a new kind of token to Sentry which are fundamentally per-organization
+tokens, but with the ability to carry meta information that tools like sentry-cli can use to
+improve the user experience.  These org level tokens can be created by anyone in the org, they
+can be given additional restrictions, and they can carry meta information such as routing
+data.  For the purpose of this document they are called **structural tokens**.
+
+## Token Format
+
+The proposed token format is to leverage JWT as serialization format.  The goals of the
+token align generally with both [Macaroons](http://macaroons.io/) and
+[Biscuit](https://www.biscuitsec.org) but unfortunately the former standard has never seen
+much attention, and the latter is pretty new, not particularly proven and very complex.
+Either system however permits adding additional restrictions to the token which make them
+a very interesting choice for the use in our pipeline.
+
+One of the benefits of having the tokens carry additional data is that the token alone has enough
+information available to route to a Sentry installation.  This means that `sentry-cli` or
+any other tool _just_ needs the token to even determine the host that the token should be
+sent against.  This benefit also applies to JWT or PASETO tokens which can be considered
+for this as well.  The RFC here thus proposes to encode this data into a regular **JWT**
+token.
+
+A serialized token is added a custom prefix `sntrys_` (sentry structure) to make
+it possible to detect it by security scrapers.  Anyone handling such a token is
+required to check for the `sntrys_` prefix and disregard it before parsing it.  This
+can also be used by the client side to detect a structural token if the client is
+interested in extracting data from the token.
+
+## Token Facts
+
+We want to encode certain information into the tokens.  The following attributes are defined:
+
+* `sentry_site`: references the target API URL that should be used.  A token will always have a
```

Comment:
Maybe specify here that this is `options['system.url-prefix']`, for reference?

Reply:
referencing this comment from @markstory here: https://github.com/getsentry/sentry/pull/50409#discussion_r1224758745

> In the future, organizations will have a couple URLs:

> sentryUrl which is our root domain and is the same as system.url-prefix. ex. sentry.io
organizationUrl which is domain for the organization's UI. ex acme.sentry.io
regionUrl Which is the domain that the organization's API endpoints are available on. ex us.sentry.io
Any API requests sent to sentryUrl will be routed to the correct region, but will suffer a latency penalty (because we're proxying the request to the region) and the request will go to the US (this can matter for EU customers).

> For this scenario, I think you'll want both the sentryUrl and regionUrl available in the JWT. There are a small number of API endpoints that are only available on sentryUrl (users, integrations, sentry apps).

So based on this, I would say we should add `sentry_url` and `sentry_region_url` fields instead of `sentry_site`?

Reply:
why is anything other than the api url needed?